zerver/lib/push_notifications.py

```diff
@@ -754,6 +769,51 @@ def handle_remove_push_notification(user_profile_id: int, message_ids: List[int]
         flags=F('flags').bitand(
             ~UserMessage.flags.active_mobile_push_notification))
 
+def encrypt_apns_payload(payload: Dict[str, Any], key: str) -> Dict[str, Any]:
+    encrypted_payload = copy.deepcopy(payload)
+    body = encrypted_payload['alert']['body']
+    encrypted_body, nonce = encrypt_data(key, body)
+    encrypted_payload['alert']['body'] = encrypted_body
+    zulip_data = encrypted_payload['custom']['zulip']
+    zulip_data.update({
+        'nonce': nonce,
+    })
+    return encrypted_payload
+
+def encrypt_gcm_payload(payload: Dict[str, Any], key: str) -> Dict[str, Any]:
+    encrypted_payload = copy.deepcopy(payload)
+    content = encrypted_payload['content']
+    encrypted_content, nonce = encrypt_data(key, content)
+    encrypted_payload['content'] = encrypted_content
+    encrypted_payload['nonce'] = nonce
+    return encrypted_payload
+
+def generate_encrypted_data(devices: List[PushDeviceToken],
+                            payload: Dict[str, Any],
+                            encrypt_payload: Callable[..., Dict[str, Any]]
+                            ) -> EncryptedData:
+    push_data = []
+    for device in devices:
+        key = device.notification_encryption_key
+        encrypted_payload = encrypt_payload(payload, key)
+        push_data.append((device, encrypted_payload))
+
+    return push_data
+
+def separate_devices(devices: List[PushDeviceToken]) -> Tuple[List[PushDeviceToken], List[PushDeviceToken]]:
```

Comment:
Maybe worth adding a comment here that we can send all the unencrypted notices to the third-party API with all the user's devices in a single request, but encrypted ones need to be one-per-device, and that's why we have this structure.